File: src/main/java/com/github/tetrisanalyzer/piece/PieceJ.java
package com.github.tetrisanalyzer.piece;

public class PieceJ extends Piece {
    @Override public byte number() { return 4; }
    @Override public char character() { return 'J'; }
    @Override protected int[] widths() { return new int[] { 3, 2, 3, 2 }; }
    @Override protected int[] heights() { return new int[] { 2, 3, 2, 3 }; }
    @Override protected PieceShape[] shapes() {
        return new PieceShape[] {
            new PieceShape(new Point(0,0), new Point(1,0), new Point(2,0), new Point(2,1)),
            new PieceShape(new Point(0,0), new Point(1,1), new Point(0,1), new Point(0,2)),
            new PieceShape(new Point(0,0), new Point(0,1), new Point(1,1), new Point(2,1)),
            new PieceShape(new Point(1,0), new Point(1,1), new Point(0,2), new Point(1,2))
        };
    }
}

File: langs/scala/src/main/scala/nu/tengstrand/tetrisanalyzer/piece/PieceJ.scala
package nu.tengstrand.tetrisanalyzer.piece

object PieceJ {
  def apply() = new PieceJ
}

class PieceJ extends Piece {
  val number = 4.toByte
  val character = 'J'
  protected val widths = Array(3, 2, 3, 2)
  protected val heights = Array(2, 3, 2, 3)
  protected val shapes = Array(
    new PieceShape(Array(Point(0,0), Point(1,0), Point(2,0), Point(2,1))),
    new PieceShape(Array(Point(0,0), Point(1,0), Point(0,1), Point(0,2))),
    new PieceShape(Array(Point(0,0), Point(0,1), Point(1,1), Point(2,1))),
    new PieceShape(Array(Point(1,0), Point(1,1), Point(0,2), Point(1,2)))
  )
}
